### ISSUE TITLE
Reproducibility with new rng structure

### ISSUE THREAD

**Issue by jkoeller** [CitrineInformatics/smlb#14]
Recently the random number generator structure was modified such that repeated calls to a learner's `.fit()` method would not yield an object that would make the same predictions. This is desirable because it allows one to determine some notion of sensitivity of predictions to the random state, for example. 

However, there are situations in which it is desirable to have reproducibility with respect to multiple calls to `.fit()`. For example, when comparing performance of a learner while varying something else about a workflow -- features, training set, etc. -- it is natural have repeated calls to `.fit()` which should yield the same result. The recent modification to the random state structure has made this more difficult. 

One solution is to make a new learner object for each configuration so that each learner only has `.fit()` called once. But naive ways to make a "copy" don't seem to work:
- `copy.copy()` does not do "sufficient copying" of the learner to achieve the desired result
- `copy.deepcopy()` of a learner leads to `RecursionError: maximum recursion depth exceeded while calling a Python object`
- `learner.__class__()` or `type(learner)()` won't work because different learners have different constructor arguments
- `sklearn.base.clone()` is sklearn's solution to estimator copying, but the smlb learners do not respect the sklearn interface.

Any suggestions about how to go about this? Is there something I'm missing?

**Comment by jkoeller** [comment]
It seems to me that fundamentally we need a way to "clone" a learner. Perhaps similar insufficiencies with copy and deepcopy explain why sklearn implemented a custom clone method.

**Comment by jkoeller** [comment]
A notebook which illustrates the issue can be found here: https://github.com/CitrineInformatics/smlb/blob/feature/new-random-structure-illustration/examples/rng%2C%20reproduciblity%20upon%20multiple%20.fit.ipynb

**Comment by bfolie** [comment]
We have two different types of behavior that are both motivated: either the individual fits _or_ the series of fits are reproducible. I see three options:
1. Figure out how to clone
2. Make "individual reproducible fits" the default option. If the user needs to sample from the space of all fits they can create a bunch of learners with different rng seeds.
3. Expose a configurable parameter that lets the user decide which type of behavior to exhibit

**Comment by bfolie** [comment]
I think it depends on the use cases. If we think both are well-motivated then I would go with option 3. If we think that "individual reproducible fits" will be the overwhelming default then I'm fine with number 2.

**Comment by jkoeller** [comment]
Agreed that both use cases are well-motivated. I actually think option 1 may be the most robust, in the sense that this "cloning" operation explicitly expresses the intent. But I don't have a good sense of what it would look like to implement that. 

With respect to 2 vs 3 I would favor 2, because (i) I do think it is more common and (ii) I don't immediately see any downside to creating a bunch of learners with different rng seeds if the desire is to sample from the space of fits.

**Comment by bfolie** [comment]
I forgot to say it explicitly, but I scratched option 1 because of its complexity. At least, it's not a good near-term solution.

@matthias-rupp what do you think?

**Comment by matthias-rupp** [comment]
A `clone` method should not be necessary.

The original design was as follows:

* Any part that utilizes pseudo-random numbers, including learners, accepts a seed (or a pseudo-random number generator, PRNG, in a defined state, which is equivalent). The behavior of that part is then deterministic given the seed. In particular, if new PRN-dependent parts are created within this part (e.g., an ensemble learner that internally trains an ensemble of models), the seed is split and each new PRN-dependent part starts with its own (deterministic) seed. This works even for distributed/parallel settings.
* To repeat the behavior of a single learner in a specific situation (training set etc.), repeatedly create the learner with the same inputs and use the same seed. This should always yield the same model.
* To repeat the behavior of a sequence of trained models, start the sequence with the same seed.

> However, there are situations in which it is desirable to have reproducibility with respect to multiple calls to .fit(). For example, when comparing performance of a learner while varying something else about a workflow -- features, training set, etc. -- it is natural have repeated calls to .fit() which should yield the same result. The recent modification to the random state structure has made this more difficult.

This setting I don't understand. If, e.g., features change, then the learned model will be different, i.e., `fit` will yield a different model. It should _not_ yield the same result because something changed. If the change does not affect the model (e.g., different validation set), creating a new learner with the same seed should yield identical results.

Let me know if I misunderstood.

**Comment by bfolie** [comment]
> If, e.g., features change, then the learned model will be different, i.e., fit will yield a different model. It should not yield the same result because something changed. If the change does not affect the model (e.g., different validation set), creating a new learner with the same seed should yield identical results.

I cannot speak to the situations Jason is envisioning, but I'm thinking about wanting to create multiple identical copies of the same response surface, for example for an optimization study. Different validation sets would be in the same category.

I suppose either behavior is achievable in either case. In one situation the user has to create multiple learners with the same seed. In the other case they have to create several learners and manually give them different seeds. When framed that way I agree that the former (the status quo) is better because it requires manually selecting fewer rng seeds.

**Comment by matthias-rupp** [comment]
I prefer the first solution (create multiple learners) because one of the design philosophies of `smlb` is that all objects are completely specified at initialization time. Re-specifying the seed would violate that. This approach allows for easy configuration from files, for example.

**Comment by jkoeller** [comment]
Hi @matthias-rupp,

> This setting I don't understand. If, e.g., features change, then the learned model will be different, i.e., `fit` will yield a different model. It should _not_ yield the same result because something changed. If the change does not affect the model (e.g., different validation set), creating a new learner with the same seed should yield identical results.

Perhaps I should clarify what I mean. I am referring to the behavior in situations like the one here: https://github.com/CitrineInformatics/smlb/blob/bd71de34dfde3f631bb0fba629bbb5a8ee3712b8/smlb/workflows/learning_curve_regression.py#L159. Imagine running this workflow with multiple training samplers and a single learner. The problem with the current implementation is that each time `.fit()` is called the learner is in a different internal state, and so the effect of the training sampler isn't the _only_ thing being varied. This means in particular that one's conclusions about the effect of different training samplers (in this example) could depend on the _order_ in which the training samplers were passed to the workflow, which seems inconsistent with the smlb philosophy.

This is more impactful when doing simulated sequential learning: the positive feedback means the outcome is highly dependent on the initial state -- so the variation in the space of fits matters more -- and the computations take a long time to run -- so just repeating it for many trials to "average out" the effect of the variation in the space of fits is inconvenient. 

The "create multiple learners with the same seed" solution would work, but then you would have to pass a different learner for each training sampler in this example. This is also fine, but would require a refactor of this workflow, among other things.

**Comment by matthias-rupp** [comment]
Hi @jkoeller 

Thanks for the explanation.

> [results could] depend on the _order_ in which the training samplers were passed to the workflow, which seems inconsistent with the smlb philosophy.

I agree. That would be inconsistent with `smlb` design.

> The "create multiple learners with the same seed" solution would work, but then you would have to pass a different learner for each training sampler in this example. This is also fine, but would require a refactor of this workflow, among other things.

This would be my suggestion. How about this pseudo-code loop?

```
for i, lcfg in enumerate(self._learner_configurations): 
    for j, training in enumerate(training_data): 
        learner = instantiate_smlb_object(lcfg)
        learner.fit(training) 
        predictions[i, j] = learner.apply(validation_data)
```

The problem is of course the instantiation. There was code for this, but unfortunately, it did not make it from my prototype into the `smlb` repository proper. 😞  However, it's not difficult: the learner configuration can be a dictionary that contains the name of the learner class and all keyword arguments. The function then simply instantiates that class with those arguments.

**Comment by jkoeller** [comment]
Ok, yes that should certainly work. This seems like a safe implementation. It is similar to the "clone" option -- a new learner for each configuration -- but requires that the user explicitly specify every aspect of the learner that is to be "cloned" through specifying the `learner_configuration`. 

Practical implementation cost aside, do you see an advantage to this as opposed to just providing a `clone` function that instantiates a new, identical learner with "fresh" internal state?
